ecl/hqlcpp/hqlttcpp.cpp

```diff
@@ -8865,6 +8865,228 @@ IHqlExpression * FilteredIndexOptimizer::createTransformed(IHqlExpression * expr
 //==============================================================================================================
 
 
+static HqlTransformerInfo DFSLayoutTransformerInfo("DFSLayoutTransformer");
+DFSLayoutTransformer::DFSLayoutTransformer(IErrorReceiver &_errs, ICodegenContextCallback * _ctxCallback, HqlCppOptions const &_options)
+: NewHqlTransformer(DFSLayoutTransformerInfo), errs(_errs), options(_options)
+{
+    ctxCallback = _ctxCallback;
+}
+
+
+IHqlExpression * DFSLayoutTransformer::createTransformed(IHqlExpression * expr)
+{
+    OwnedHqlExpr transformed = NewHqlTransformer::createTransformed(expr);
+    int filenameIdx = 0;
+    int recordIdx = 0;
+    int transformIdx = -1;
+    switch (transformed->getOperator())
+    {
+    default:
+        return transformed.getClear();
+    case no_table:
+        filenameIdx = 0;
+        recordIdx = 1;
+        break;
+    case no_keyindex:
+        throwUnexpected();  // They have already been translated into no_newkeyindex by HqlTreeNormalizer::transformKeyIndex
+        break;
+    case no_newkeyindex:
+        // Args are flatfile, record, transform, name. Offset of first payload field indicated by value of __payload_Atom attribute
+        filenameIdx = 3;
+        transformIdx = 2;
+        recordIdx = 1;
+        break;
+    }
+    LinkedHqlExpr ds = transformed;
+    IHqlExpression * filename = ds->queryChild(filenameIdx);
+    bool translate = getBoolAttribute(ds, lookupAtom, options.translateDFSlayouts && filename->isConstant());
+    if (translate)
+    {
+        ECLlocation where(ds);
+        if (!filename->isConstant())
+            errs.reportError(ERR_EXPECTED_CONST, "LOOKUP attribute requires a constant filename", str(where.sourcePath), where.lineno, where.column, where.position);
+        StringBuffer fileName;
+        filename->queryValue()->getStringValue(fileName);
```

Comment:
isConstant() isn't sufficient to ensure that filename->queryValue() is set, E.g. it could be concat of two constants..  That is only true if the expression has been folded (and even then there are some other constant expressions like ALL).  That may not have been done by this point.  
Also getStringValue(fileName, filename) is safer.